File: DAA/assignment3/py_progs/strassen.py
# Write a Python program to implement Strassen’s matrix multiplication.

import numpy as np


def strassen_multiply(A, B):
    n = A.shape[0]

    # Base case: if the matrices are 1x1
    if n == 1:
        return A * B

    # Split matrices into quadrants
    A11 = A[:n//2, :n//2]
    A12 = A[:n//2, n//2:]
    A21 = A[n//2:, :n//2]
    A22 = A[n//2:, n//2:]

    B11 = B[:n//2, :n//2]
    B12 = B[:n//2, n//2:]
    B21 = B[n//2:, :n//2]
    B22 = B[n//2:, n//2:]

    # Recursively compute products of smaller matrices
    P1 = strassen_multiply(A11 + A22, B11 + B22)
    P2 = strassen_multiply(A21 + A22, B11)
    P3 = strassen_multiply(A11, B12 - B22)
    P4 = strassen_multiply(A22, B21 - B11)
    P5 = strassen_multiply(A11 + A12, B22)
    P6 = strassen_multiply(A21 - A11, B11 + B12)
    P7 = strassen_multiply(A12 - A22, B21 + B22)

    # Compute the quadrants of the product matrix
    C11 = P1 + P4 - P5 + P7
    C12 = P3 + P5
    C21 = P2 + P4
    C22 = P1 - P2 + P3 + P6

    # Combine the quadrants into a single matrix and return
    C = np.vstack((np.hstack((C11, C12)), np.hstack((C21, C22))))
    return C


if __name__ == "__main__":
    n = 4
    A = [[i+j for j in range(n)] for i in range(n)]
    B = [[2*A[i][j] for j in range(n)] for i in range(n)]
    print("Strassen's Matrix Multiplication in Python")
    print("A matrix =")
    print(A)
    print("B matrix =")
    print(B)
    A = np.array(A)
    B = np.array(B)
    C = strassen_multiply(A, B)
    C = C.tolist()
    print("A*B matrix =")
    print(C)

# OUTPUT
# Strassen's Matrix Multiplication in Python
# A matrix =
# [[0, 1, 2, 3], [1, 2, 3, 4], [2, 3, 4, 5], [3, 4, 5, 6]]
# B matrix =
# [[0, 2, 4, 6], [2, 4, 6, 8], [4, 6, 8, 10], [6, 8, 10, 12]]
# A*B matrix =
# [[28, 40, 52, 64], [40, 60, 80, 100], [52, 80, 108, 136], [64, 100, 136, 172]]

File: DAA/assignment3/c_progs/strassen.c
// Write a C program to implement Strassen’s matrix multiplication.

#include <stdio.h>
#include <stdlib.h>

void strassen(int n, int A[][n], int B[][n], int C[][n]);
void add(int n, int A[][n], int B[][n], int C[][n]);
void subtract(int n, int A[][n], int B[][n], int C[][n]);

int main()
{
    int n = 4, i, j;
    int A[n][n], B[n][n], C[n][n];

    for (i = 0; i < n; i++)
    {
        for (j = 0; j < n; j++)
        {
            A[i][j] = i + j;
            B[i][j] = 2 * A[i][j];
        }
    }

    printf("Strassen's Matrix Multiplication in C\n", n, n);
    printf("Order of both matrices: %d * %d\n", n, n);
    strassen(n, A, B, C);

    printf("\nA matrix=\n");
    for (i = 0; i < n; i++)
    {
        for (j = 0; j < n; j++)
            printf("%d\t", A[i][j]);
        printf("\n");
    }

    printf("\nB matrix=\n");
    for (i = 0; i < n; i++)
    {
        for (j = 0; j < n; j++)
            printf("%d\t", B[i][j]);
        printf("\n");
    }

    printf("\nAxB matrix=\n");
    for (i = 0; i < n; i++)
    {
        for (j = 0; j < n; j++)
            printf("%d\t", C[i][j]);
        printf("\n");
    }

    return 0;
}

void strassen(int n, int A[][n], int B[][n], int C[][n])
{
    if (n == 1)
    {
        C[0][0] = A[0][0] * B[0][0];
        return;
    }

    int i, j;

    int A11[n / 2][n / 2], A12[n / 2][n / 2], A21[n / 2][n / 2], A22[n / 2][n / 2];
    int B11[n / 2][n / 2], B12[n / 2][n / 2], B21[n / 2][n / 2], B22[n / 2][n / 2];
    int C11[n / 2][n / 2], C12[n / 2][n / 2], C21[n / 2][n / 2], C22[n / 2][n / 2];
    int P1[n / 2][n / 2], P2[n / 2][n / 2], P3[n / 2][n / 2], P4[n / 2][n / 2], P5[n / 2][n / 2], P6[n / 2][n / 2], P7[n / 2][n / 2];
    int temp1[n / 2][n / 2], temp2[n / 2][n / 2];

    // Divide A into 4 submatrices
    for (i = 0; i < n / 2; i++)
        for (j = 0; j < n / 2; j++)
            A11[i][j] = A[i][j];
    for (i = 0; i < n / 2; i++)
        for (j = n / 2; j < n; j++)
            A12[i][j - n / 2] = A[i][j];
    for (i = n / 2; i < n; i++)
        for (j = 0; j < n / 2; j++)
            A21[i - n / 2][j] = A[i][j];
    for (i = n / 2; i < n; i++)
        for (j = n / 2; j < n; j++)
            A22[i - n / 2][j - n / 2] = A[i][j];

    // Divide B into 4 submatrices
    for (i = 0; i < n / 2; i++)
        for (j = 0; j < n / 2; j++)
            B11[i][j] = B[i][j];
    for (i = 0; i < n / 2; i++)
        for (j = n / 2; j < n; j++)
            B12[i][j - n / 2] = B[i][j];
    for (i = n / 2; i < n; i++)
        for (j = 0; j < n / 2; j++)
            B21[i - n / 2][j] = B[i][j];
    for (i = n / 2; i < n; i++)
        for (j = n / 2; j < n; j++)
            B22[i - n / 2][j - n / 2] = B[i][j];

    // Calculate the 7 products
    add(n / 2, A11, A22, temp1);
    add(n / 2, B11, B22, temp2);
    strassen(n / 2, temp1, temp2, P1);

    add(n / 2, A21, A22, temp1);
    strassen(n / 2, temp1, B11, P2);

    subtract(n / 2, B12, B22, temp1);
    strassen(n / 2, A11, temp1, P3);

    subtract(n / 2, B21, B11, temp1);
    strassen(n / 2, A22, temp1, P4);

    add(n / 2, A11, A12, temp1);
    strassen(n / 2, temp1, B22, P5);

    subtract(n / 2, A21, A11, temp1);
    add(n / 2, B11, B12, temp2);
    strassen(n / 2, temp1, temp2, P6);

    subtract(n / 2, A12, A22, temp1);
    add(n / 2, B21, B22, temp2);
    strassen(n / 2, temp1, temp2, P7);

    // Calculate the 4 quadrants of C using the products
    add(n / 2, P1, P4, temp1);
    subtract(n / 2, temp1, P5, temp2);
    add(n / 2, temp2, P7, C11);

    add(n / 2, P3, P5, C12);

    add(n / 2, P2, P4, C21);

    add(n / 2, P1, P3, temp1);
    subtract(n / 2, temp1, P2, temp2);
    add(n / 2, temp2, P6, C22);

    // Combine the 4 quadrants of C into one matrix
    for (i = 0; i < n / 2; i++)
        for (j = 0; j < n / 2; j++)
            C[i][j] = C11[i][j];
    for (i = 0; i < n / 2; i++)
        for (j = n / 2; j < n; j++)
            C[i][j] = C12[i][j - n / 2];
    for (i = n / 2; i < n; i++)
        for (j = 0; j < n / 2; j++)
            C[i][j] = C21[i - n / 2][j];
    for (i = n / 2; i < n; i++)
        for (j = n / 2; j < n; j++)
            C[i][j] = C22[i - n / 2][j - n / 2];
}

void add(int n, int A[][n], int B[][n], int C[][n])
{
    int i, j;
    for (i = 0; i < n; i++)
        for (j = 0; j < n; j++)
            C[i][j] = A[i][j] + B[i][j];
}

void subtract(int n, int A[][n], int B[][n], int C[][n])
{
    int i, j;
    for (i = 0; i < n; i++)
        for (j = 0; j < n; j++)
            C[i][j] = A[i][j] - B[i][j];
}

// OUTPUT
// Strassen's Matrix Multiplication
// Order of both matrices: 4 * 4

// A matrix=
// 0       1       2       3
// 1       2       3       4
// 2       3       4       5
// 3       4       5       6

// B matrix=
// 0       2       4       6
// 2       4       6       8
// 4       6       8       10
// 6       8       10      12

// AxB matrix=
// 28      40      52      64
// 40      60      80      100
// 52      80      108     136
// 64      100     136     172